README.md
# Consul HTTP API Ansible Module

[![Latest Version](http://img.shields.io/github/release/mtchavez/consul-ansible-module.svg?style=flat-square)](https://github.com/mtchavez/consul-ansible-module/releases)
[![Build Status](https://travis-ci.org/mtchavez/consul-ansible-module.svg?branch=master)](https://travis-ci.org/mtchavez/consul-ansible-module)

An Ansible module to interact with consul's API from your playbooks and roles.

## API Endpoints

* [ACL](#acl)
* [Events](#events)
* [Key/Value](#keyvalue)
* [Session](#session)
* [Status](#status)

### [ACL](#acl)

#### Usage

```yaml
- name: Create ACL with defaults
  consul_acl:
    action: create
    token: "master-token"

- name: Update ACL
  consul_acl:
    action: update
    acl_id: "existing-acl-id"
    name: "updated-acl-name"
    token: "master-token"

- name: List ACL
  consul_acl:
    action: list
    token: "master-token"

- name: Get ACL replication
  consul_acl:
    action: replication
    token: "master-token"

- name: Get ACL info
  consul_acl:
    action: info
    acl_id: "asdf-1234-asdf-1234"
    token: "master-token"

- name: Destroy ACL
  consul_acl:
    action: destroy
    acl_id: "asdf-1234-asdf-1234"
    token: "master-token"

- name: Clone ACL
  consul_acl:
    action: clone
    acl_id: "asdf-1234-asdf-1234"
    token: "master-token"
```

#### Documentation

```yaml
---
module: consul_acl
version_added: "1.9"
author: Chavez
short_description: Interact with Consul ACL API
description:
   - Use Consul ACL API in your playbooks and roles
options:
  acl_type:
    description:
      - Type of ACL
    required: false
  action:
    description:
      - One of [create]
    required: true
  dc:
    desription:
      - The datacenter to use
    required: false
    default: dc1
  host:
    description:
      - Consul host
    required: true
    default: 127.0.0.1
  name:
    description:
      - Name of event to fire
    required: false
  port:
    description:
      - Consul API port
    required: true
  rules:
    description:
      - ACL rules to set or update
    required: false
  token:
    description:
      - ACL token to use with requests
    required: false
  version:
    description:
      - Consul API version
    required: true
    default: v1
```


### [Events](#events)

#### Usage

Examples

```yaml
- name: Event new
  consul_event:
    action: fire
    name: deploy
  register: new_event

- name: Event new
  consul_event:
    action: list
  register: all_events

```

#### Documentation

```yaml
---
module: consul_events
version_added: "1.9"
author: Chavez
short_description: Interact with Consul Event API
description:
   - Use Consul Event API in your playbooks and roles
options:
  action:
    description:
      - One of [leader, peers]
    required: true
  dc:
    desription:
      - The datacenter to use
    required: false
    default: dc1
  host:
    description:
      - Consul host
    required: true
    default: 127.0.0.1
  name:
    description:
      - Name of event to fire
    required: false
  node:
    description:
      - Node query filter for event fire
    required: false
  port:
    description:
      - Consul API port
    required: true
  service:
```

### [Key/Value](#keyvalue)

#### Usage

Examples

```yaml
- name: PUT a value for a key
  consul_kv:
    action: put
    key: foo
    value: bar

- name: PUT value with flag
  consul_kv:
    action: put
    key: bar/baz/bizzle
    value: shizzle
    flags: 23

- name: GET key for PUT with check and set
  consul_kv:
    action: get
    key: bar/baz/bizzle
  register: bizzle

- name: PUT check and set
  consul_kv:
    action: put
    key: bar/baz/bizzle
    value: "no shizzle"
    cas: "{{ item.ModifyIndex|int }}"
  with_items: bizzle.value

- name: PUT with session
  consul_kv:
    action: put
    key: razzle/acquired
    value: true
    acquire: "some-valid-session"

- name: PUT with session release
  consul_kv:
    action: put
    key: razzle/acquired
    value: true
    release: "some-valid-session"

- name: GET a value for a key
  consul_kv:
    action: get
    key: foo/bar/baz

- name: GET keys for prefix
  consul_kv:
    action: get
    key: bar
    keys: true
  register: bar_keys

- name: GET keys up to separator
  consul_kv:
    action: get
    key: bar/
    keys: true
    separator: "/"
  register: separator_keys

- name: DELETE a key
  consul_kv:
    action: delete
    key: foo/tmp

- name: DELETE a directory recursively
  consul_kv:
    action: delete
    key: foo/bar
    recurse: true
```

#### Documentation

```yaml
module: consul_kv
version_added: "1.9"
author: Chavez
short_description: Interact with Consul K/V API
description:
   - Use Consul K/V API in your playbooks and roles
options:
  acquire:
    - description:
      - Session to use for PUT requests
    required: false

    description:
      - HTTP verb [GET, PUT, DELETE]
    required: true
  dc:
    desription:
      - The datacenter to use
    required: false
    default: dc1
  cas:
    description:
      - Check and set parameter
    require: false
  flags:
    description:
      - Opaque flag to set as metadata for a key
    require: false
  host:
    description:
      - Consul host
    required: true
    default: 127.0.0.1
  key:
    description:
      - Key to interact with in K/V store
    required: true
  keys:
    description:
      - Return keys on a GET request for a given path
    required: false
    default: False
  port:
    description:
      - Consul API port
    required: true
    default: 8500
  recurse:
    description:
      - Recurse flag for DELETE or GET actions
    required: false
    default: False
  release:
    - description:
      - Session to release for PUT requests
    required: false
  separator:
    description:
      - Separator to use when listing keys for a GET
    required: false
  value:
    description:
      - Value to set when adding or updating a key
    required: false
  version:
    description:
      - Consul API version
    required: true
    default: v1
```

### [Session](#session)

#### Usage

Examples
```yaml
- name: Session create
  consul_session:
    action: create
    ttl: 15s
  register: session

- name: Get session info
  consul_session:
    action: info
    session: "{{ session.value.ID }}"
  register: session_info

- name: Session create with all params
  consul_session:
    action: create
    lock_delay: 25s
    node: "{{ item.Node }}"
    behavior: delete
    checks: ""
    ttl: 25s
  register: session_params
  with_items: "{{ session_info.value }}"

- name: Get session params info
  consul_session:
    action: info
    session: "{{ item.value.ID }}"
  register: session_params_info
  with_items: "{{ session_params.results }}"

- name: Get session renew
  consul_session:
    action: renew
    session: "{{ session.value.ID }}"
  register: session_renew

- name: List sessions
  consul_session:
    action: list
  register: all_sessions

- name: All sessions for a node
  consul_session:
    action: node
    node: "{{ item.Node }}"
  register: node_sessions
  with_items: "{{ session_info.value }}"

- name: Destroy session
  consul_session:
    action: destroy
    session: "{{ item }}"
  with_items:
    - "{{ session.value.ID }}"
    - "{{ session_params.results[0].value.ID }}"
```

#### Documentation
```yaml
module: consul_session
version_added: "1.9"
author: Chavez
short_description: Interact with Consul Sessions API
description:
   - Use Consul Sessions API in your playbooks and roles
options:
  action:
    description:
      - API session action [create, destroy, info, node, list, renew]
    required: true
  behavior:
    description:
      - Controls when the session is invalidated [release, delete]
    require: false
  checks:
    description:
      - List of associated health checks comma separated "foo,bar,baz"
    required: false
  dc:
    desription:
      - The datacenter to use
    required: false
    default: dc1
  host:
    description:
      - Consul host
    required: true
    default: 127.0.0.1
  lock_delay:
    description:
      - Time to delay the lock of the session
    require: false
  node:
    description:
      - Node name to set on create
    required: false
  port:
    description:
      - Consul API port
    required: true
  session:
    description:
      - Consul session to interact with
    require: false
  ttl:
    description:
      - Session TTL
    required: false
  version:
    description:
      - Consul API version
    required: true
    default: v1
```

### [Status](#status)

#### Usage

Examples

```yaml
- name: Get leader
  consul_status:
    action: leader
  register: leader

- name: Get peers
  consul_status:
    action: peers
  register: peers
```

#### Documentation
```yaml
module: consul_status
version_added: "1.9"
author: Chavez
short_description: Interact with Consul Status API
description:
   - Use Consul Status API in your playbooks and roles
options:
  action:
    description:
      - One of [leader, peers]
    required: true
  dc:
    desription:
      - The datacenter to use
    required: false
    default: dc1
  host:
    description:
      - Consul host
    required: true
    default: 127.0.0.1
  port:
    description:
      - Consul API port
    required: true
  version:
    description:
      - Consul API version
    required: true
    default: v1

# informational: requirements for nodes
requirements: [ ]
```

## Testing

You will need to be running a consul server locally to run the test playbook.

Run consul with
  * DC1: `consul agent --server --bootstrap-expect=1 --data-dir=ci/consul/dc1/ --log-level=debug --config-file=ci/consul/dc1/config.json`
  * DC2: `consul agent --server --bootstrap-expect=1 --config-file=ci/consul/dc2/config.json --data-dir=ci/consul/dc2 --log-level=debug`

Then you can run the test playbook with `ansible-playbook -i ./hosts test-consul.yml`

## TODO

- ACL API
  - [x]  `/v1/acl/create`
  - [x]  `/v1/acl/update`
  - [x]  `/v1/acl/destroy/<id>`
  - [x]  `/v1/acl/info/<id>`
  - [x]  `/v1/acl/clone/<id>`
  - [x]  `/v1/acl/list`
  - [x]  `/v1/acl/replication`
- Agent API
  - [ ] `/v1/agent/checks`
  - [ ] `/v1/agent/services`
  - [ ] `/v1/agent/members`
  - [ ] `/v1/agent/self`
  - [ ] `/v1/agent/maintenance`
  - [ ] `/v1/agent/join/<address>`
  - [ ] `/v1/agent/force-leave/<node>`
  - [ ] `/v1/agent/check/register`
  - [ ] `/v1/agent/check/deregister/<checkID>`
  - [ ] `/v1/agent/check/pass/<checkID>`
  - [ ] `/v1/agent/check/warn/<checkID>`
  - [ ] `/v1/agent/check/fail/<checkID>`
  - [ ] `/v1/agent/check/update/<checkID>`
  - [ ] `/v1/agent/service/register`
  - [ ] `/v1/agent/service/deregister/<serviceID>`
  - [ ] `/v1/agent/service/maintenance/<serviceID>`
- Catalog API
  - [ ] `/v1/catalog/register`
  - [ ] `/v1/catalog/deregister`
  - [ ] `/v1/catalog/datacenters`
  - [ ] `/v1/catalog/nodes`
  - [ ] `/v1/catalog/services`
  - [ ] `/v1/catalog/service/<service>`
  - [ ] `/v1/catalog/node/<node>`
- Coordinate API
  - [ ] ` /v1/coordinate/datacenters`
  - [ ] ` /v1/coordinate/nodes`
- Events API
  - [x] fire
  - [x] list
- Health Checks API
  - [ ] `/v1/health/node/<node>`
  - [ ] `/v1/health/checks/<service>`
  - [ ] `/v1/health/service/<service>`
  - [ ] `/v1/health/state/<state>`
- Key/Value API
  - [x] GET
  - [x] PUT
  - [x] DELETE
  - [x] Session acquire PUT
  - [x] Session release PUT
- Prepared Queries API
  - [ ] `/v1/query`
  - [ ] `/v1/query/<query`
  - [ ] `/v1/query/<query or name>/execute`
  - [ ] `/v1/query/<query or name>/explain`
- Session API
  - [x] create
  - [x] destroy
  - [x] info
  - [x] node
  - [x] list
  - [x] renew
- Status API
  - [x] leader
  - [x] peers


CHANGELOG.md
# v0.5.0

- acl: create added
- acl: update added
- acl: list added
- acl: replication query added
- acl: info added
- acl: destroy added
- acl: clone added

# v0.4.0

- Update existing modules (k/v, session, status) to ansible standards
- event: fire
- event: list

# v0.3.0

* status: leader
* status: peers

# v0.2.0

* session: Destroy
* session: Info
* session: Renew
* session: List
* session: Node sessions

# v0.1.0

* kv: Datacenter param added
* kv: Recurse DELETE of directory
* kv: GET all keys for a given prefix
* kv: DELETE with check and set
* kv: GET keys up to separator
* kv: PUT flags param added
* kv: PUT check and set
* kv: PUT with session acquire
* kv: PUT with session release
* session: Create

# v0.0.1

* kv: GET value for key
* kv: PUT value for key
* kv: DELETE key


library/consul_kv.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# Copyright 2015 Chavez <chavez@somewhere-cool.com>
#
# This file is part of Ansible
#
# Ansible is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Ansible is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.

import base64
import json
import string
import urllib

from collections import OrderedDict


DOCUMENTATION = '''
---
module: consul_kv
version_added: "1.9"
author: Chavez
short_description: Interact with Consul K/V API
description:
   - Use Consul K/V API in your playbooks and roles
options:
  acquire:
    - description:
      - Session to use for PUT requests
    required: false
  action:
    description:
      - HTTP verb [GET, PUT, DELETE]
    required: true
  dc:
    desription:
      - The datacenter to use
    required: false
    default: dc1
  cas:
    description:
      - Check and set parameter
    require: false
  flags:
    description:
      - Opaque flag to set as metadata for a key
    require: false
  host:
    description:
      - Consul host
    required: true
    default: 127.0.0.1
  key:
    description:
      - Key to interact with in K/V store
    required: true
  keys:
    description:
      - Return keys on a GET request for a given path
    required: false
    default: False
  port:
    description:
      - Consul API port
    required: true
  recurse:
    description:
      - Recurse flag for DELETE or GET actions
    required: false
    default: False
  release:
    - description:
      - Session to release for PUT requests
    required: false
  separator:
    description:
      - Separator to use when listing keys for a GET
    required: false
  token:
    description:
      - ACL token to use with requests
  value:
    description:
      - Value to set when adding or updating a key
    required: false
  version:
    description:
      - Consul API version
    required: true
    default: v1

# informational: requirements for nodes
requirements: [ urllib ]
'''

EXAMPLES = '''
# PUT a value for a key
- consul_kv: action=put key=foo value=bar

# PUT value with flag
- consul_kv: action=put key=bar/baz/bizzle value="shizzle" flags=23

# GET key for PUT with check and set
- consul_kv: action=get key=bar/baz/bizzle
  register: bizzle

# PUT check and set
- consul_kv: action=put key=bar/baz/bizzle value="no shizzle" cas={{item.ModifyIndex|int}}
  with_items: bizzle.value

# PUT with session
- consul_kv: action=put key=razzle/acquired value="true" acquire="some-valid-session"

# PUT with session release
- consul_kv: action=put key=razzle/acquired value="true" acquire="some-valid-session release="some-valid-session" "

# GET a value for a key
- consul_kv: action=get key=foo/bar/baz

# GET keys for prefix
- consul_kv: action=get key=bar keys=true
  register: bar_keys

# GET keys up to separator
- consul_kv: action=get key=bar/ keys=true separator='/'
  register: separator_keys

# DELETE a key
- consul_kv: action=delete key=foo/tmp

# DELETE a directory recursively
- consul_kv: action=delete key=foo/bar recurse=true
'''

#
# Module execution.
#


class ConsulKV(object):

    ALLOWED_ACTIONS = ['GET', 'PUT', 'DELETE']
    GET, PUT, DELETE = ALLOWED_ACTIONS

    def __init__(self, module):
        """Takes an AnsibleModule object to set up Consul K/V interaction"""
        self.module = module
        self.acquire = module.params.get('acquire', None)
        self.action = string.upper(module.params.get('action', ''))
        self.cas = module.params.get('cas', None)
        self.dc = module.params.get('dc', 'dc1')
        self.flags = module.params.get('flags', None)
        self.host = module.params.get('host', '127.0.0.1')
        self.key = module.params.get('key', '')
        self.keys = module.params.get('keys', False)
        self.port = module.params.get('port', 8500)
        self.recurse = module.params.get('recurse', False)
        self.release = module.params.get('release', None)
        self.separator = module.params.get('separator', None)
        self.token = module.params.get('token', None)
        self.value = module.params.get('value', '')
        self.version = module.params.get('version', 'v1')
        self._build_url()

    def run_cmd(self):
        self.validate()
        self._make_api_call()

    def validate(self):
        # Check action is allowed
        if not self.action or self.action not in self.ALLOWED_ACTIONS:
            self.module.fail_json(msg='Action is required and must be one of GET, PUT, DELETE')
        # A key is required for any call so make sure one exists
        if not self.key:
            self.module.fail_json(msg='A key is required to interact with the k/v API')
        # Validate action being used
        # ie self._validate_get(), self._validate_put(), self._validate_delete()
        getattr(self, "_validate_%s" % string.lower(self.action))

    def _build_url(self):
        self.api_url = "http://%s:%s/%s/kv/%s" % (self.host, self.port, self.version, self.key)

    def _validate_get(self):
        pass

    def _validate_put(self):
        if not self.value:
            self.module.fail_json(msg='A value is required when using PUT')

    def _validate_delete(self):
        pass

    def _make_api_call(self):
        self._setup_request()

        try:
            (response, info) = fetch_url(module, self.api_url, data=self.value, method=self.action)
        except urllib2.URLError, e:
            self.module.fail_json(msg="API call failed: {}, info: {}".format(str(e), info))

        try:
            response_body = response.read()
            self._handle_response(response, response_body)
        except AttributeError, e:
            self.module.fail_json(msg="Parsing response failed: {}".format(str(e), info))


    def _query_params(self):
        params = OrderedDict({})
        if self.dc != 'dc1':
            params['dc'] = self.dc
        if self.token:
            params['token'] = self.token
        if self.action == self.DELETE and self.recurse:
            params['recurse'] = 'true'
        if self.action in [self.DELETE, self.PUT] and self.cas:
            params['cas'] = self.cas
        if self.action == self.GET and self.keys:
            if self.separator:
                params['separator'] = self.separator
            params['keys'] = 'true'
        if self.action == self.PUT:
            if self.flags:
                params['flags'] = self.flags
            if self.acquire:
                params['acquire'] = self.acquire
            if self.release:
                params['release'] = self.release
        return params

    def _setup_request(self):
        params = urllib.urlencode(self._query_params())
        if params:
            self.api_url = self.api_url + '?' + params

    def _handle_response(self, response, response_body):
        if self.action == self.PUT and response_body == 'true':
            self.module.exit_json(changed=True, succeeded=True, key=self.key, value=self.value)
        elif self.action == self.DELETE and response.getcode() == 200:
            self.module.exit_json(changed=True, succeeded=True, key=self.key, deleted=True)
        elif self.action == self.GET:
            parsed_response = json.loads(response_body)
            # Decode values
            for obj in parsed_response:
                # When doing a GET for only keys the objects will not
                # be a dict with metadata but only a list of string key values
                if isinstance(obj, dict):
                    obj['Value'] = base64.decodestring(obj.get('Value', ''))
            self.module.exit_json(changed=True, succeeded=True, key=self.key, value=parsed_response)
        else:
            self.module.fail_json(msg="Failed %s with a %i for key: %s because %s" % (self.action, response.getcode(), self.key, response_body))


def main():
    global module
    module = AnsibleModule(
        argument_spec=dict(
            acquire=dict(require=False),
            action=dict(required=True),
            cas=dict(require=False, type='int'),
            dc=dict(required=False, default='dc1'),
            flags=dict(require=False, type='int'),
            host=dict(required=False, default="127.0.0.1"),
            key=dict(required=True),
            keys=dict(require=False, default=False, type='bool'),
            port=dict(require=False, default=8500),
            recurse=dict(require=False, default=False, type='bool'),
            release=dict(require=False),
            separator=dict(require=False),
            token=dict(required=False, default=None),
            value=dict(required=False),
            version=dict(required=False, default='v1'),
        ),
        supports_check_mode=True
    )

    # If we're in check mode, just exit pretending like we succeeded
    if module.check_mode:
        module.exit_json(changed=False)

    consulkv = ConsulKV(module)
    consulkv.run_cmd()


# import module snippets
from ansible.module_utils.basic import *
from ansible.module_utils.urls import *

if __name__ == '__main__':
    main()


library/consul_status.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# Copyright 2015 Chavez <chavez@somewhere-cool.com>
#
# This file is part of Ansible
#
# Ansible is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Ansible is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.

import json
import string


DOCUMENTATION = '''
---
module: consul_status
version_added: "1.9"
author: Chavez
short_description: Interact with Consul Status API
description:
   - Use Consul Status API in your playbooks and roles
options:
  action:
    description:
      - One of [leader, peers]
    required: true
  dc:
    desription:
      - The datacenter to use
    required: false
    default: dc1
  host:
    description:
      - Consul host
    required: true
    default: 127.0.0.1
  port:
    description:
      - Consul API port
    required: true
  version:
    description:
      - Consul API version
    required: true
    default: v1

# informational: requirements for nodes
requirements: []
'''

EXAMPLES = '''
# Get leader
- consul_status: action=leader

# Get peers
- consul_status: action=peers
'''

#
# Module execution.
#


class ConsulStatus(object):

    ALLOWED_ACTIONS = ['leader', 'peers']
    LEADER, PEERS = ALLOWED_ACTIONS

    def __init__(self, module):
        """Takes an AnsibleModule object to set up Consul Status interaction"""
        self.module = module
        self.action = string.lower(module.params.get('action', ''))
        self.dc = module.params.get('dc', 'dc1')
        self.host = module.params.get('host', '127.0.0.1')
        self.port = module.params.get('port', 8500)
        self.version = module.params.get('version', 'v1')
        self._build_url()

    def run_cmd(self):
        self._make_api_call()

    def validate(self):
        # Check action is allowed
        if not self.action or self.action not in self.ALLOWED_ACTIONS:
            self.module.fail_json(msg='Action is required and must be one of %r' % self.ALLOWED_ACTIONS)
        # Validate action being used
        # ie self._validate_create(), self._validate_destroy(), self._validate_info()
        getattr(self, "_validate_%s" % self.action)

    def _build_url(self):
        self.api_url = "http://%s:%s/%s/status/%s" % (self.host, self.port, self.version, self.action)

    def _make_api_call(self):
        self._setup_request()

        try:
            (response, info) = fetch_url(module, self.api_url)
        except Exception, e:
            self.module.fail_json(msg="API call ({}) failed: {}".format(self.api_url, str(e)))

        response_body = response.read()
        self._handle_response(response, response_body)

    def _setup_request(self):
        # Add dc param if not the default
        if self.dc != 'dc1':
            self.api_url = self.api_url + '?dc=%s' % self.dc

    def _handle_response(self, response, response_body):
        code = response.getcode()
        if code != 200:
            self.module.fail_json(msg="Failed with code %i and response %s" % (code, response_body))
        else:
            try:
                parsed_response = json.loads(response_body)
            except:
                parsed_response = ''
            self.module.exit_json(changed=True, succeeded=True, value=parsed_response)


def main():
    global module
    module = AnsibleModule(
        argument_spec=dict(
            action=dict(required=True),
            dc=dict(required=False, default='dc1'),
            host=dict(required=False, default='127.0.0.1'),
            port=dict(require=False, default=8500),
            version=dict(required=False, default='v1'),
        ),
        supports_check_mode=True
    )

    # If we're in check mode, just exit pretending like we succeeded
    if module.check_mode:
        module.exit_json(changed=False)

    consul_status = ConsulStatus(module)
    consul_status.run_cmd()


# import module snippets
from ansible.module_utils.basic import *
from ansible.module_utils.urls import *

if __name__ == '__main__':
    main()


library/consul_session.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# Copyright 2015 Chavez <chavez@somewhere-cool.com>
#
# This file is part of Ansible
#
# Ansible is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Ansible is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.

import json
import string
import urllib

from collections import OrderedDict


DOCUMENTATION = '''
---
module: consul_session
version_added: "1.9"
author: Chavez
short_description: Interact with Consul Sessions API
description:
   - Use Consul Sessions API in your playbooks and roles
options:
  action:
    description:
      - API session action [create, destroy, info, node, list, renew]
    required: true
  behavior:
    description:
      - Controls when the session is invalidated [release, delete]
    require: false
  checks:
    description:
      - List of associated health checks comma separated "foo,bar,baz"
    required: false
  dc:
    desription:
      - The datacenter to use
    required: false
    default: dc1
  host:
    description:
      - Consul host
    required: true
    default: 127.0.0.1
  lock_delay:
    description:
      - Time to delay the lock of the session
    require: false
  node:
    description:
      - Node name to set on create
    required: false
  port:
    description:
      - Consul API port
    required: true
  session:
    description:
      - Consul session to interact with
    require: false
  token:
    description:
      - ACL token to use with requests
  ttl:
    description:
      - Session TTL
    required: false
  version:
    description:
      - Consul API version
    required: true
    default: v1

# informational: requirements for nodes
requirements: []
'''

EXAMPLES = '''
# Session create
- consul_session: action=create

# Session destroy
- consul_session: action=destroy session="some-valid-session"

# Get session info
- consul_session: action=info session="some-valid-session"

# Renew session
- consul_session: action=renew session="some-valid-session"

# List sessions
- consul_session: action=list

# List sessions
- consul_session: action=list
  register: all_sessions

# All sessions for a node
- consul_session: action=node node="node-foo"
'''

#
# Module execution.
#


class ConsulSession(object):

    ALLOWED_ACTIONS = ['create', 'destroy', 'info', 'node', 'list', 'renew']
    CREATE, DESTROY, INFO, NODE, LIST, RENEW = ALLOWED_ACTIONS

    PUT_ACTIONS = [CREATE, DESTROY, RENEW]
    GET_ACTIONS = [INFO, NODE, LIST]

    DEFAULT_CHECKS = ['serfHealth']

    def __init__(self, module):
        """Takes an AnsibleModule object to set up Consul Session interaction"""
        self.module = module
        self.action = string.lower(module.params.get('action', ''))
        self.behavior = module.params.get('behavior', 'release')
        self.checks = module.params.get('checks', self.DEFAULT_CHECKS[0])
        self.dc = module.params.get('dc', 'dc1')
        self.host = module.params.get('host', '127.0.0.1')
        self.lock_delay = module.params.get('lock_delay', '15s')
        self.node = module.params.get('node', '')
        self.port = module.params.get('port', 8500)
        self.session = module.params.get('session', '')
        self.token = module.params.get('token', '')
        self.ttl = module.params.get('ttl', '15s')
        self.version = module.params.get('version', 'v1')
        self.params = OrderedDict({})
        self.req_data = ''
        self.checks = str(self.checks).split(',')
        self._build_url()

    def run_cmd(self):
        self.validate()
        self._make_api_call()

    def validate(self):
        # Check action is allowed
        if not self.action or self.action not in self.ALLOWED_ACTIONS:
            self.module.fail_json(msg='Action is required and must be one of %r' % self.ALLOWED_ACTIONS)
        # Validate action being used
        # ie self._validate_create(), self._validate_destroy(), self._validate_info()
        getattr(self, "_validate_%s" % self.action)

    def _build_url(self):
        self.api_url = "http://%s:%s/%s/session/%s" % (self.host, self.port, self.version, self.action)
        if self.action in [self.DESTROY, self.INFO, self.RENEW] and self.session:
            self.api_url += '/%s' % self.session
        if self.action == self.NODE:
            self.api_url += '/%s' % self.node

    def _validate_create(self):
        pass

    def _validate_destroy(self):
        if not self.session:
            module.fail_json(msg="Destroy requires a session")

    def _validate_info(self):
        if not self.session:
            module.fail_json(msg="Info requires a session")

    def _validate_renew(self):
        if not self.session:
            module.fail_json(msg="Renew requires a session")

    def _validate_node(self):
        pass

    def _validate_list(self):
        pass

    def _make_api_call(self):
        self._setup_request()

        try:
            (response, info) = fetch_url(module, self.api_url, data=self.req_data, method=self._http_verb_for_action())
        except Exception, e:
            self.module.fail_json(msg="API call ({}) failed: {}".format(self.api_url, str(e)))

        try:
            response_body = response.read()
            self._handle_response(response, response_body)
        except AttributeError, e:
            self.module.fail_json(msg="Parsing response failed: {}, info: {}".format(str(e), info))

    def _add_create_params(self):
        valid_params = {
            "lock_delay": "LockDelay",
            "name": "Name",
            "node": "Node",
            "checks": "Checks",
            "behavior": "Behavior",
            "ttl": "TTL"
        }
        for param, name in valid_params.iteritems():
            if hasattr(self, param) and getattr(self, param):
                self.params[name] = getattr(self, param)
        # Ensure the default checks exist
        if self.DEFAULT_CHECKS[0] not in self.params['Checks']:
            self.params['Checks'] += self.DEFAULT_CHECKS
        self.params['Checks'] = filter(None, self.params['Checks'])

    def _setup_request(self):
        params = urllib.urlencode(self._query_params())
        if params:
            self.api_url = self.api_url + '?' + params
        # Add params for CREATE
        # Will set self.params dictionary to be encoded
        if self.action == self.CREATE:
            self._add_create_params()

        if self.action in self.PUT_ACTIONS:
            if self.params:
                self.req_data = json.dumps(self.params)

    def _query_params(self):
        params = OrderedDict({})
        if self.dc != 'dc1':
            params['dc'] = self.dc
        if self.token:
            params['token'] = self.token
        return params

    def _http_verb_for_action(self):
        if self.action in self.PUT_ACTIONS:
            return 'PUT'
        return 'GET'

    def _handle_response(self, response, response_body):
        code = response.getcode()
        if code != 200:
            self.module.fail_json(msg="Failed with code %i for session %s with response %s" % (code, self.action, response_body))
        else:
            try:
                parsed_response = json.loads(response_body)
            except:
                parsed_response = ''
            self.module.exit_json(changed=True, succeeded=True, value=parsed_response)


def main():
    global module
    module = AnsibleModule(
        argument_spec=dict(
            action=dict(required=True),
            behavior=dict(required=False, default='release'),
            checks=dict(required=False, default=ConsulSession.DEFAULT_CHECKS[0]),
            dc=dict(required=False, default='dc1'),
            host=dict(required=False, default='127.0.0.1'),
            lock_delay=dict(require=False, default='15s'),
            node=dict(required=False),
            port=dict(require=False, default=8500),
            session=dict(require=False),
            token=dict(required=False, default=None),
            ttl=dict(required=False, default='15s'),
            version=dict(required=False, default='v1'),
        ),
        supports_check_mode=True
    )

    # If we're in check mode, just exit pretending like we succeeded
    if module.check_mode:
        module.exit_json(changed=False)

    consul_session = ConsulSession(module)
    consul_session.run_cmd()


# import module snippets
from ansible.module_utils.basic import *
from ansible.module_utils.urls import *

if __name__ == '__main__':
    main()
